components/feature/push/src/main/java/mozilla/components/feature/push/AutoPushFeature.kt

```diff
@@ -0,0 +1,325 @@
+ *              val feature = AutoPushFeature(context, service, configuration)
+ *              PushProvider.install(push)
+ *          }
+ *     }
+ * </code>
+ *
+ * Subscription information for each registered [PushType] can be retrieved asynchronously:
+ *
+ * <code>
+ *    feature.registerForSubscriptions(object: Bus.Observer<PushType, String> {
+ *        override fun onSubscriptionAvailable(subscription: AutoPushSubscription) { }
+ *    })
+ *    feature.subscribeForType(PushType.Services)
+ * </code>
+ *
+ * Push messages for each registered [PushType] can also be retrieved asynchronously:
+ *
+ * <code>
+ *     feature.registerForPushMessages(object: Bus.Observer<PushType, String> {
+ *        override fun onEvent(message: String) { }
+ *     })
+ * </code>
+ *
+ */
+@Suppress("TooManyFunctions")
+class AutoPushFeature(
+    private val context: Context,
+    private val service: PushService,
+    config: PushConfig,
+    coroutineContext: CoroutineContext = Executors.newSingleThreadExecutor().asCoroutineDispatcher(),
+    private val connection: PushConnection = RustPushConnection(
+        senderId = config.senderId,
+        serverHost = config.serverHost,
+        socketProtocol = config.protocol,
+        serviceType = config.serviceType,
+        databasePath = File(context.filesDir, DB_NAME).canonicalPath
+    )
+) : PushProcessor {
+
+    private val logger = Logger("AutoPushFeature")
+    private val subscriptionObservers: Observable<PushSubscriptionObserver> = ObserverRegistry()
+    private val messageObserverBus: Bus<PushType, String> = MessageBus()
+    private val prefToken = preferences(context).getString(PREF_TOKEN, null)
+
+    internal var job: Job = SupervisorJob()
+    private val scope = CoroutineScope(coroutineContext) + job
+
+    init {
+        // If we have a token, initialize the rust component first.
+        prefToken?.let { token ->
+            runBlocking(coroutineContext) {
+                connection.updateToken(token)
+            }
+        }
+    }
+
+    /**
+     * Starts the push service provided.
+     */
+    override fun initialize() { service.start(context) }
+
+    /**
+     * Register to receive push subscriptions when requested or when they have been re-registered.
+     */
+    fun registerForSubscriptions(
+        observer: PushSubscriptionObserver,
+        owner: LifecycleOwner,
+        autoPause: Boolean
+    ) = subscriptionObservers.register(observer, owner, autoPause)
+
+    /**
+     * Register to receive push messages for the associated [PushType].
+     */
+    fun registerForPushMessages(
+        type: PushType,
+        observer: Bus.Observer<PushType, String>,
+        owner: LifecycleOwner,
+        autoPause: Boolean
+    ) = messageObserverBus.register(type, observer, owner, autoPause)
+
+    /**
+     * Un-subscribes from all push message channels. This should only be done on an account logout or app data deletion.
+     */
+    override fun shutdown() {
```

Comment:
nit: group `overrides` together at the top of the class

Reply:
Rename this method. `unsubcribeAll` is a better name. Current name implies you call this during a regular app lifecycle, and that this method is oppose to `initialize`.